File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for m in range(len(nums)):
            if m > 0 and nums[m] == nums[m-1]:
                continue
            l, r = m + 1, len(nums) - 1
            while l < r:
                tmp = nums[l] + nums[m] + nums[r]
                if tmp == 0:
                    res.append([nums[m], nums[l], nums[r]]) 
                    l += 1
                    while l < r and nums[l] == nums[l-1]:
                        l += 1
                    r -= 1
                elif tmp < 0:
                    l += 1
                else:
                    r -= 1
        return res

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.cpp
#define vi vector<int>
#define pb push_back

class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        if (candidates.empty()) return {};
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vi subset;
        dfs(0, 0, candidates, subset, res, target);
        return res;
    }

private:
    void dfs(int i, int currentSum, vector<int>& candidates, vi& subset, vector<vector<int>>& res, int target) {
        if (currentSum == target) {
            res.pb(subset);
            return;
        }
        if (i >= candidates.size() || currentSum > target) {
            return;
        }

        // take nums[i]
        subset.pb(candidates[i]);
        dfs(i + 1, currentSum + candidates[i], candidates, subset, res, target);
        subset.pop_back();

        // skip nums[i] and dups!
        while (i + 1 < candidates.size() && candidates[i] == candidates[i + 1]) {
            i++;
        }
        dfs(i + 1, currentSum, candidates, subset, res, target);
    }
};